### PR TITLE
GUI refactoring to integrate with Keycloak

### PR DESCRIPTION
**Pull request by cfrancisco**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This PR introduces a code refactoring that aims to use the new keycloak service as an authorization schema. Below, we have listed all these changes:
- Removed code for the forget password and update password methods;
- Removed all code related to user management;
- Removed all code related to the management of groups;
- Removed the permissions management page;
- Updated permission system to be retrocompatible with the previous system;
- Implementing a new data model for user data;
- Updated all endpoints with new URLs (to proxy's endpoint); 
- In node-red's code, all references to JWT/auth-token have been removed;
- Updated methods for storing authentication and removed getToken functions;
- Node-red: Fixed some broken URLs;
- LocalStorage should not need a polyfill anymore, the browser compatibility can be checked in:
https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage 

Several minor changes have been made as well, including: 
- I18n: fixed few bugs to get the user language;
- Applied Eslint in a wide variety of components;
- A bug was fixed when Devices' Map was requested without positions;
- From now on, instead of combining terms services and tenants, the term Tenant will now be used everywhere;

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)
This PR is related to dojot/dojot#2110

* **Other information**:

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -7,7 +7,7 @@ WORKDIR /data
 ARG DOJOT_VERSION='undefined'
 
 #It is used to define the API URL
-ARG BASE_URL='/'
+ARG BASE_URL=''
```
Comment (mprevide) [inline Dockerfile:10]: Are you sure about that?

File: src/js/comms/groups/GroupManager.js [hunk 2]
```diff
@@ -1,25 +1,25 @@
-import { baseURL } from 'Src/config';
```
Comment (mprevide) [inline src/js/comms/groups/GroupManager.js:1]: You can delete this file

File: src/js/views/flows/edit.js [hunk 3]
```diff
@@ -180,7 +180,7 @@ class FlowCanvas extends Component {
                     <div id="editor-stack" />
 
                     <div id="palette" style={this.cannotEdit ? { display: 'none' } : {}}>
-                        <img src={`${baseURL}flows/red/images/spin.svg`} className="palette-spinner hide" />
+                        <img src={`${PROXY_URL}flows/red/images/spin.svg`} className="palette-spinner hide" />
```
Comment (mprevide) [inline src/js/views/flows/edit.js:183]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 4]
```diff
@@ -4054,41 +4056,41 @@ RED.tabs = (function() {
...
-        str: {value:"str",label:"string",icon:`${baseURL}flows/red/images/typedInput/az.png`},
-        num: {value:"num",label:"number",icon:`${baseURL}flows/red/images/typedInput/09.png`,validate:/^[+-]?[0-9]*\.?[0-9]*([eE][-+]?[0-9]+)?$/},
-        bool: {value:"bool",label:"boolean",icon:`${baseURL}flows/red/images/typedInput/bool.png`,options:["true","false"]},
-        json: {value:"json",label:"JSON",icon:`${baseURL}flows/red/images/typedInput/json.png`, validate: function(v) { try{JSON.parse(v);return true;}catch(e){return false;}}},
-        re: {value:"re",label:"regular expression",icon:`${baseURL}flows/red/images/typedInput/re.png`},
-        date: {value:"date",label:"timestamp",hasValue:false},
+        msg: { value: "msg", label: "msg.", validate: RED.utils.validatePropertyExpression },
+        flow: { value: "flow", label: "flow.", validate: RED.utils.validatePropertyExpression },
+        global: { value: "global", label: "global.", validate: RED.utils.validatePropertyExpression },
+        str: { value: "str", label: "string", icon: `${PROXY_URL}flows/red/images/typedInput/az.png` },
```
Comment (mprevide) [inline src/js/views/flows/red.js:4064]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 5]
```diff
@@ -4054,41 +4056,41 @@ RED.tabs = (function() {
...
+        num: { value: "num", label: "number", icon: `${PROXY_URL}flows/red/images/typedInput/09.png`, validate: /^[+-]?[0-9]*\.?[0-9]*([eE][-+]?[0-9]+)?$/ },
```
Comment (mprevide) [inline src/js/views/flows/red.js:4065]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 6]
```diff
@@ -4054,41 +4056,41 @@ RED.tabs = (function() {
...
+        bool: { value: "bool", label: "boolean", icon: `${PROXY_URL}flows/red/images/typedInput/bool.png`, options: ["true", "false"] },
```
Comment (mprevide) [inline src/js/views/flows/red.js:4066]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 7]
```diff
@@ -4054,41 +4056,41 @@ RED.tabs = (function() {
...
+        json: { value: "json", label: "JSON", icon: `${PROXY_URL}flows/red/images/typedInput/json.png`, validate: function (v) { try { JSON.parse(v); return true; } catch (e) { return false; } } },
```
Comment (mprevide) [inline src/js/views/flows/red.js:4067]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 8]
```diff
@@ -4054,41 +4056,41 @@ RED.tabs = (function() {
...
+        re: { value: "re", label: "regular expression", icon: `${PROXY_URL}flows/red/images/typedInput/re.png` },
```
Comment (mprevide) [inline src/js/views/flows/red.js:4068]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 9]
```diff
@@ -4054,41 +4056,41 @@ RED.tabs = (function() {
...
+        date: { value: "date", label: "timestamp", hasValue: false },
         jsonata: {
             value: "jsonata",
             label: "expression",
-            icon: `${baseURL}flows/red/images/typedInput/expr.png`,
-            validate: function(v) { try{jsonata(v);return true;}catch(e){return false;}},
-            expand:function() {
+            icon: `${PROXY_URL}flows/red/images/typedInput/expr.png`,
```
Comment (mprevide) [inline src/js/views/flows/red.js:4073]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 10]
```diff
@@ -4512,64 +4514,64 @@ RED.actions = (function() {
...
+        59: 186,
+        61: 187,
+        173: 189
     }
 
     function init() {
 
         $.getJSON({
-            url: `${baseURL}flows/red/keymap.json`,
+            url: `${PROXY_URL}flows/red/keymap.json`,
```
Comment (mprevide) [inline src/js/views/flows/red.js:4563]: Remove proxy use here

File: src/js/views/flows/red.js [hunk 11]
```diff
@@ -4512,64 +4514,64 @@ RED.actions = (function() {
...
             type: 'GET',
             headers: {
                 "authorization": `Bearer ${util.getToken()}`,
```
Comment (mprevide) [inline src/js/views/flows/red.js:4566]: Leave this line

File: src/js/views/flows/red.js [hunk 12]
```diff
@@ -5048,20 +5050,20 @@ RED.workspaces = (function() {
...
+        active: function () {
             return activeWorkspace
         },
-        show: function(id) {
+        show: function (id) {
             if (!workspace_tabs.contains(id)) {
                 var sf = RED.nodes.subflow(id);
                 if (sf) {
-                    addWorkspace({type:"subflow",id:id,icon:`${baseURL}flows/red/images/subflow_tab.png`,label:sf.name, closeable: true});
+                    addWorkspace({ type: "subflow", id: id, icon: `${PROXY_URL}flows/red/images/subflow_tab.png`, label: sf.name, closeable: true });
```
Comment (mprevide) [inline src/js/views/flows/red.js:5066]: Remove proxy use

File: src/js/views/flows/red.js [hunk 13]
```diff
@@ -6920,35 +6924,35 @@ RED.view = (function() {
...
+                        .attr("fill-opacity", "0.05")
+                        .attr("height", function (d) { return Math.min(50, d.h - 4); });
 
                     var icon = icon_group.append("image")
-                        .attr("xlink:href",`${baseURL}flows/icons/${+d._def.icon}`)
-                        .attr("class","node_icon")
-                        .attr("x",0)
-                        .attr("width","30")
-                        .attr("height","30");
+                        .attr("xlink:href", `${PROXY_URL}flows/icons/${+d._def.icon}`)
```
Comment (mprevide) [inline src/js/views/flows/red.js:6940]: Remove proxy use

File: src/js/views/flows/red.js [hunk 14]
```diff
@@ -6964,11 +6968,11 @@ RED.view = (function() {
                     //}
 
                     var img = new Image();
-                    img.src = `${baseURL}flows/icons/${d._def.icon}`;
-                    img.onload = function() {
-                        icon.attr("width",Math.min(img.width,30));
-                        icon.attr("height",Math.min(img.height,30));
-                        icon.attr("x",15-Math.min(img.width,30)/2);
+                    img.src = `${PROXY_URL}flows/icons/${d._def.icon}`;
```
Comment (mprevide) [inline src/js/views/flows/red.js:6971]: Remove proxy use

File: src/js/views/flows/red.js [hunk 15]
```diff
@@ -6977,69 +6981,69 @@ RED.view = (function() {
...
-                        .attr("x",20).attr("y",9);
+                        .attr("class", "node_status_label")
+                        .attr("x", 20).attr("y", 9);
                 }
                 //node.append("circle").attr({"class":"centerDot","cx":0,"cy":0,"r":5});
 
                 //node.append("path").attr("class","node_error").attr("d","M 3,-3 l 10,0 l -5,-8 z");
-                node.append("image").attr("class","node_error hidden").attr("xlink:href",`${baseURL}flows/icons/node-error.png`).attr("x",0).attr("y",-6).attr("width",10).attr("height",9);
-                node.append("image").attr("class","node_changed hidden").attr("xlink:href",`${baseURL}flows/icons/node-changed.png`).attr("x",12).attr("y",-6).attr("width",10).attr("height",10);
+                node.append("image").attr("class", "node_error hidden").attr("xlink:href", `${PROXY_URL}flows/icons/node-error.png`).attr("x", 0).attr("y", -6).attr("width", 10).attr("height", 9);
```
Comment (mprevide) [inline src/js/views/flows/red.js:7008]: Remove proxy use

File: src/js/views/flows/red.js [hunk 16]
```diff
@@ -6977,69 +6981,69 @@ RED.view = (function() {
...
+                node.append("image").attr("class", "node_changed hidden").attr("xlink:href", `${PROXY_URL}flows/icons/node-changed.png`).attr("x", 12).attr("y", -6).attr("width", 10).attr("height", 10);
```
Comment (mprevide) [inline src/js/views/flows/red.js:7009]: Remove proxy use

File: src/js/views/flows/red.js [hunk 17]
```diff
@@ -7122,61 +7126,61 @@ RED.view = (function() {
...
+                                } catch (err) {
+                                    console.log("icon", err);
                                     icon_url = "arrow-in.png";
                                 }
                             } else {
                                 icon_url = d._def.icon;
                             }
-                            if (`${baseURL}flows/icons/${icon_url}` !== current_url) {
-                                icon.attr("xlink:href",`${baseURL}flows/icons/${icon_url}`);
+                            if (`${PROXY_URL}flows/icons/${icon_url}` !== current_url) {
```
Comment (mprevide) [inline src/js/views/flows/red.js:7136]: Remove proxy use

File: src/js/views/flows/red.js [hunk 18]
```diff
@@ -7122,61 +7126,61 @@ RED.view = (function() {
...
+                                icon.attr("xlink:href", `${PROXY_URL}flows/icons/${icon_url}`);
                                 var img = new Image();
-                                img.src = `${baseURL}flows/icons/${d._def.icon}`;
-                                img.onload = function() {
-                                    icon.attr("width",Math.min(img.width,30));
-                                    icon.attr("height",Math.min(img.height,30));
-                                    icon.attr("x",15-Math.min(img.width,30)/2);
+                                img.src = `${PROXY_URL}flows/icons/${d._def.icon}`;
```
Comment (mprevide) [inline src/js/views/flows/red.js:7139]: Remove proxy use

File: src/js/views/flows/red.js [hunk 19]
```diff
@@ -7798,76 +7802,76 @@ RED.palette = (function() {
...
                     icon_url = (typeof def.icon === "function" ? def.icon.call({}) : def.icon);
-                } catch(err) {
-                    console.log("Definition error: "+nt+".icon",err);
+                } catch (err) {
+                    console.log("Definition error: " + nt + ".icon", err);
                 }
-                var iconContainer = $('<div/>',{class:"palette_icon_container"+(def.align==="right"?" palette_icon_container_right":"")}).appendTo(d);
-                $('<div/>',{class:"palette_icon",style:`background-image: url(${baseURL}flows/icons/${icon_url})`}).appendTo(iconContainer);
+                var iconContainer = $('<div/>', { class: "palette_icon_container" + (def.align === "right" ? " palette_icon_container_right" : "") }).appendTo(d);
+                $('<div/>', { class: "palette_icon", style: `background-image: url(${PROXY_URL}flows/icons/${icon_url})` }).appendTo(iconContainer);
```
Comment (mprevide) [inline src/js/views/flows/red.js:7874]: Remove proxy use

File: src/js/views/flows/red.js [hunk 20]
```diff
@@ -10966,58 +10970,58 @@ RED.typeSearch = (function() {
...
+                        console.log("Definition error: " + object.type + ".icon", err);
                     }
                 }
-                nodeDiv.css('backgroundColor',colour);
+                nodeDiv.css('backgroundColor', colour);
 
-                var iconContainer = $('<div/>',{class:"palette_icon_container"}).appendTo(nodeDiv);
-                $('<div/>',{class:"palette_icon",style:`background-image: url(${baseURL}flows/icons/${icon_url})`}).appendTo(iconContainer);
+                var iconContainer = $('<div/>', { class: "palette_icon_container" }).appendTo(nodeDiv);
+                $('<div/>', { class: "palette_icon", style: `background-image: url(${PROXY_URL}flows/icons/${icon_url})` }).appendTo(iconContainer);
```
Comment (mprevide) [inline src/js/views/flows/red.js:11008]: Remove proxy use

File: src/js/views/flows/red.js [hunk 21]
```diff
@@ -55,27 +55,27 @@ RED.events = (function() {
...
                     defaultNS: 'editor',
                     backend: {
-                        loadPath: function(lngs, ns) {
-                            return `${baseURL}flows/locales/${ns}`;
+                        loadPath: function (lngs, ns) {
+                            return `${PROXY_URL}flows/locales/${ns}`;
                         },
                         withCredentials: true,
-                        customHeaders: {'Authorization': accessToken},
+                        customHeaders: { 'Authorization': accessToken },
```
Comment (mprevide) [inline src/js/views/flows/red.js:72]: Remove this line?

File: src/js/views/flows/red.js [hunk 22]
```diff
@@ -156,35 +156,35 @@ RED.settings = (function () {
         var accessTokenMatch = /[?&]access_token=(.*?)(?:$|&)/.exec(window.location.search);
         if (accessTokenMatch) {
             var accessToken = accessTokenMatch[1];
-            RED.settings.set("auth-tokens",{access_token: accessToken});
+            RED.settings.set("auth-tokens", { access_token: accessToken });
```
Comment (mprevide) [inline src/js/views/flows/red.js:159]: Remove this line?

File: src/js/views/flows/red.js [hunk 23]
```diff
@@ -156,35 +156,35 @@ RED.settings = (function () {
...
 
         $.ajaxSetup({
-            beforeSend: function(jqXHR,settings) {
+            beforeSend: function (jqXHR, settings) {
                 // Only attach auth header for requests to relative paths
                 if (!/^\s*(https?:|\/|\.)/.test(settings.url)) {
                     var auth_tokens = RED.settings.get("auth-tokens");
                     if (auth_tokens) {
-                        jqXHR.setRequestHeader("Authorization","Bearer "+auth_tokens.access_token);
+                        jqXHR.setRequestHeader("Authorization", "Bearer " + auth_tokens.access_token);
```
Comment (mprevide) [inline src/js/views/flows/red.js:169]: Remove this line?

File: src/js/views/flows/red.js [hunk 24]
```diff
@@ -156,35 +156,35 @@ RED.settings = (function () {
...
-            RED.settings.set("auth-tokens",{access_token: accessToken});
+            RED.settings.set("auth-tokens", { access_token: accessToken });
             window.location.search = "";
         }
 
         $.ajaxSetup({
-            beforeSend: function(jqXHR,settings) {
+            beforeSend: function (jqXHR, settings) {
                 // Only attach auth header for requests to relative paths
                 if (!/^\s*(https?:|\/|\.)/.test(settings.url)) {
```
Comment (mprevide) [inline src/js/views/flows/red.js:166]: where does `settings.url` come from? she must have to go through the proxy

File: src/js/views/flows/red.js [hunk 25]
```diff
@@ -193,34 +193,34 @@ RED.settings = (function () {
                 // console.log("Node-RED: " + data.version);
                 done();
             },
-            error: function(jqXHR,textStatus,errorThrown) {
+            error: function (jqXHR, textStatus, errorThrown) {
                 if (jqXHR.status === 401) {
                     if (/[?&]access_token=(.*?)(?:$|&)/.test(window.location.search)) {
```
Comment (mprevide) [inline src/js/views/flows/red.js:198]: Remove this line?

File: src/js/views/flows/red.js (hunk 11 above)
Comment (mprevide) [reply src/js/views/flows/red.js:4566]: Token checks are ignored in these cases: https://github.com/dojot/flowbroker/blob/7d7e133fb3f817bd709f21c127d669e8c2a7c320/orchestrator/auth.js#L55

File: Dockerfile (hunk 1 above)
Comment (cfrancisco) [reply Dockerfile:10]: Yes... Using only '/' creates an odd request. The slash will be added afterward in the config.js.

File: src/js/comms/groups/GroupManager.js (hunk 2 above)
Comment (cfrancisco) [reply src/js/comms/groups/GroupManager.js:1]: Thanks! Removed.

File: src/js/views/flows/edit.js (hunk 3 above)
Comment (cfrancisco) [reply src/js/views/flows/edit.js:183]: Updated!

File: src/js/views/flows/red.js (hunk 4 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4064]: Updated!

File: src/js/views/flows/red.js (hunk 5 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4065]: Updated!

File: src/js/views/flows/red.js (hunk 6 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4066]: Updated!

File: src/js/views/flows/red.js (hunk 7 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4067]: Updated!

File: src/js/views/flows/red.js (hunk 8 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4068]: Updated!

File: src/js/views/flows/red.js (hunk 9 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4073]: Updated!

File: src/js/views/flows/red.js (hunk 10 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4563]: Updated!

File: src/js/views/flows/red.js (hunk 12 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:5066]: Updated!

File: src/js/views/flows/red.js (hunk 11 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:4566]: Line removed.

File: src/js/views/flows/red.js (hunk 13 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:6940]: Updated.

File: src/js/views/flows/red.js (hunk 14 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:6971]: Updated.

File: src/js/views/flows/red.js (hunk 15 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:7008]: Updated.

File: src/js/views/flows/red.js (hunk 16 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:7009]: Updated.

File: src/js/views/flows/red.js (hunk 17 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:7136]: Updated.

File: src/js/views/flows/red.js (hunk 18 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:7139]: Updated.

File: src/js/views/flows/red.js (hunk 20 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:11008]: Updated.

File: src/js/views/flows/red.js (hunk 21 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:72]: Removed!

File: src/js/views/flows/red.js (hunk 22 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:159]: Removed!

File: src/js/views/flows/red.js (hunk 23 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:169]: Removed!

File: src/js/views/flows/red.js (hunk 24 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:166]: You raise an excellent point... I will investigate.

File: src/js/views/flows/red.js (hunk 25 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:198]: Removed!

File: src/js/views/flows/red.js [hunk 26]
```diff
@@ -156,35 +156,35 @@ RED.settings = (function () {
...
                 "Accept": "application/json"
             },
             dataType: "json",
             cache: false,
             //  url: 'http://localhost:1880/settings',
-            url: `${baseURL}flows/settings`,
+            url: `${PROXY_URL}flows/settings`,
             success: function (data) {
                 setProperties(data);
                 if (!RED.settings.user || RED.settings.user.anonymous) {
```
[the remark this answers is not in the document]
Comment (cfrancisco) [reply src/js/views/flows/red.js:190]: Mmm... Not sure, I will check it out.

[the remark this answers is not in the document]
Comment (cfrancisco) [reply src/js/views/flows/red.js:190]: In fact, no. This isn't being used anymore, so I've removed it.

File: src/js/views/flows/red.js (hunk 24 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js:166]: This code is no longer useful, I'll remove it.

**Review by luisfelipefurlan — APPROVED** [review]

**Comment by mprevide** [comment]
- Try not to use force pushes on a PR, it gets in the way of the review, then you can squash when you finish the PR
- Next time make the code improvements in a separate PR, when there are few improvements I don't see a problem being in the same file, but as there are MANY, it is very difficult to review.

File: src/js/views/flows/red.js (hunk 19 above)
Comment (cfrancisco) [reply src/js/views/flows/red.js]: Updated!

**Review by mprevide — DISMISSED** [review]

**Review by mprevide — DISMISSED** [review]

**Review by mprevide — DISMISSED** [review]

**Review by luisfelipefurlan — APPROVED** [review]

**Pull request closed without merging**